README.md
moQ web chat
==============

What is it for?
--------------

moQ is (to be) a web based messaging client where two people can securely send and receive content from their web browser. Content can be embedded like images, sounds, YouTube videos and more. Chat logs are stored on the users' machines.

How do I use it?
--------------

I think that users should enter a passphrase, and if two people enter the same passphrase they are connected to the chat. I don't know if this will be the case after we develop it.

What can it do?
--------------

moQ is for...

- Sharing messages securely from the web browser
- Watching videos and other content together (proposed idea)
- Sharing images and sounds embedded into the chat (with lightbox functionality)
- It could be used for group chat rooms (proposed idea)

[[unfinished]]


changelog.txt
[moQ Changelog]


#####V_1.00#####
This will contain the features of version 1;
when there is a working version of the system.
(we are not anywhere near version 1 yet)

server.js
var express = require('express');
var io = require('socket.io');
var server = require('http');
var path = require('path');
var app = express()
  , server = require('http').createServer(app)
  , io = io.listen(server);
displaynames = [];
// Give the client access to the 'public' folder.
app.use(express.static(path.join(__dirname, 'public')));
// This is the server port.
server.listen(3000);
// When the socket connects...
io.sockets.on('connection', function(socket){
    // When a new user connects:
    socket.on('new_user', function(data, callback){
        // Make sure display name doesnt exist already.
        if(displaynames.indexOf(data) != -1){
            callback(false);
        }else{
            callback(true);
            // Put the displayname in the socket.
            socket.displayname = data;
            displaynames.push(socket.displayname);
            update_displaynames();
        }
    });
    function update_displaynames(){
        io.sockets.emit('displaynames', displaynames);
    }
    // When a message is sent.
    socket.on('msg_send', function(data){
        // Send message to everyone except me.
        socket.broadcast.emit('msg_new', {msg: data, nick: socket.displayname});
    });
    // When a user disconnects:
    socket.on('disconnect', function(data){
        if(!socket.displayname) return;
        displaynames.splice(displaynames.indexOf(socket.displayname), 1);
        update_displaynames();
    });
});

public/moq_chat.js
console.log('moQ chat processor');
// Put the version number here.
document.getElementById('verid').innerHTML="0.01";
// History is the chat log area.
// Textinput is the box that you type in to.
// Displayname is the client username.
// Submitbutton is the submit button from the input area.
var history = document.getElementById('history'),
textinput = document.getElementById('textinput'),
displayname = document.getElementById('displayname'),
submitbutton = document.getElementById('submitbutton');
// Focus the caret to the text input box, so the user can type right away!
textinput.focus();
// Clear the history box, just incase?
history.innerHTML="";
// this gets called whenever the user presses something on their
// keyboard. 
function run(e) {
        switch(e.keyCode)
        {
            // When enter button is pressed, append and process the line
            // and reset the input box to empty.
            case 13:
            e.preventDefault();
            send_text();
            break;
            // Disable up arrow key (interferes with input box).
            case 38:
            e.preventDefault();
            break;
            // Disable down arrow key (interferes with input box).
            case 40:
            e.preventDefault();
            break;
            default:
            // If text message greater than X characters, disable input.
            // ctrl f 'max message size' for explanation.
            if (textinput.value.length >= 255){
                e.preventDefault();
            }
            break;
        }
}
// This function takes the line that the user has typed in the input box
// and places it into the chat history box!
function append_line(line, system, user){
    if (line.length){ 
        // There are user messages and system messages, which can be styled
        // and differentiated in the css.
        if(system == 1){
            // If a user string is passed:
            if (user != 0){
                var new_content = document.createElement('span');
                new_content.innerHTML = "<span class=\"displayname\">"+user+"</span><span class=\"user\">"+line+"</span>";
                history.appendChild(new_content);
            }else{
                // New content is added using appendChild, as content does not
                // have to be deleted and reloaded this way.
                var new_content = document.createElement('span');
                new_content.innerHTML = "<span class=\"displayname\">"+localStorage.displayname+"</span><span class=\"user\">"+line+"</span>";
                history.appendChild(new_content);
            }
        }else if(system == 0){
            var new_content = document.createElement('span');
            new_content.innerHTML = "<span class=\"system\">"+line+"</span>";
            history.appendChild(new_content);
        }else if(system == 2){
            // This is for appending multiple lines for the user message.
            var new_content = document.createElement('span');
            new_content.innerHTML = "<span class=\"user\">"+line+"</span>";
            history.appendChild(new_content);    
        }else if(system == 3){
            // This is a hidden message.
            return;
        }
    }
    // Save the chat log into storage.
    // This is done every time a line is appended.
    localStorage.log=history.innerHTML;
    // Scroll to the bottom of the history when a new message is appended.
    history.scrollTop = history.scrollHeight;
}
// This function sends text from the input box to the history box
// It is used by enter key listener and the submit button.
function send_text(){
    data = pre_process(textinput.value);
    // Send the value of the text to the server.
    socket.emit('msg_send', data);
    // Place the value of the text input box to the history box.
    // This puts the line into the chat on the client side.
    // This probably needs cleaning up (TODO).
    append_line(data, 1, document.getElementById('displayname').value);
    // Run the value through the processor.
    process(data);
    // Reset the text input box to a blank state.
    textinput.value="";
}
// When the settings button is pressed, this function toggles the
// settings menu from the top.
function settings_toggle(){
    if(document.getElementById('settings').style.top != "50%"){
        document.getElementById('settings').style.top="50%";
    }else{
        document.getElementById('settings').style.top="-100%";
    }
}
// This function stores the settings when the user updates them
// and also updates setting specific things in moQ.
function update_settings(){
    // Set the username to the one we have in storage (if it exists).
    if(typeof(Storage)!=="undefined"){
        if (localStorage.displayname){
            localStorage.displayname=displayname.value;
            // Emit the display name change to the server.
            socket.emit('new_user',
                document.getElementById('displayname').value,
                function(data){
                    if(data){
                        // If a new username is supplied:   
                    }else{
                        // If the name is taken:
                        document.getElementById('name_error').innerHTML='Name taken';
                    }
                });
            // Also set the display name to chat history.
            document.getElementById('displayname').value=localStorage.displayname;
        }
    }
    // Every time settings is updated, go through these colours and update them.
    document.getElementById('dncolour').style.color=document.getElementById('dncolour').innerHTML;
    document.getElementById('displayname').style.color=document.getElementById('dncolour').innerHTML;
    document.getElementById('textcolour').style.color=document.getElementById('textcolour').innerHTML;
    // Select every element with user class.
    var user_colour = document.getElementsByClassName('user');
    // Go through every element with the class and set the text colour to the one in settings.
    for (var i=user_colour.length; i--;) {
        user_colour[i].style.color=document.getElementById('textcolour').innerHTML;
    }
    // Select every element with displayname class.
    var displayname_color = document.getElementsByClassName('displayname');
    // Go through every element with the class and set the text colour to the one in settings.
    for (var i=displayname_color.length; i--;) {
        displayname_color[i].style.color=document.getElementById('dncolour').innerHTML;
    }
}
// This function captures any clicks made and redirects focus to
// the text input box. Users can still select text and click links
// and stuff, this just makes it less hassle to type stuff.
document.addEventListener('click', function(e) {
    // At the moment, it takes focus when it should not, for
    // example when the user is entering things in the settings
    // page. This needs to be fixed. (TODO)
    //textinput.focus();
}, false);

// Start with the history scrolled to the bottom.
setTimeout(function(){history.scrollTop = history.scrollHeight;},200);


public/moq_network.js
// Connect socket
var socket = io.connect();
var data = "";
// When a new message is received...
socket.on('msg_new', function(data){
    // Make sure the data is not empty.
    if(data != -1){
        // Append the data to the chat history.
        // Put the nick into the append.
        append_line(data.msg, 1, data.nick);
        // Show message 'toast' on title.
        document.title=data.nick+": "+data.msg;
        // Process the data.
        process(data.msg);
        // Check to make sure this is getting called correctly
        console.log("received data: "+data.msg);
        console.log("received data: "+data.nick);
    }
});
// When displaynames array is updated:
socket.on('displaynames', function(data){
    var content = "";
    for(i=0;i<data.length;i++){
        content+=data[i]+', ';
    }
    document.getElementById('user').innerHTML=content;
});
// This function stores the chat in localStorage, so that
// the chat history is retained when the page is refreshed.
// It needs to separate the chat logs by what rooms they are in
// and probably needs some server request to make sure that
// the data was not lost - this still has a lot of work to be done.
// (TODO)
function text_log()
{
    // Make sure there is storage
    if(typeof(Storage)!=="undefined"){
        if (localStorage.log){
            // Put history contents back in its' DIV.
            history.innerHTML=localStorage.log;
        }
        else
        {
            // If it is a brand new chat - this message is displayed.
            localStorage.log="<span class=\"system\">New chat with %user%</span>";
        }
        // Do we have a display name stored?
        if(localStorage.displayname){
            // Set the display name to the one in storage.
            displayname.innerHTML=localStorage.displayname;
        }else{
            // They don't have a name.
            localStorage.displayname="Unnamed";
        }
    }else{
        // Browser does not support localStorage. A server request would
        // be done here, possibly. Server could store the logs in a 
        // database, if this things turns out that way. (TODO)
        history.innerHTML="Logs could not be loaded with this browser.";
    }
}
// Execute the text log function on page load.
text_log()


public/moq_process.js
// This function processes each line as it is appended,
// which is used for the system messages and most likely
// for embedding content (when that functionality is added).
// (TODO)
// I also want to be able to have the commands only locally
// displayed, for example if someone types clear, it should
// not be sent to everyone as a text message.
// (TODO)
function process(line){
    // Split the lines by the spaces
    var fw = line.split(" ");
    // This switch case selects the first word of the split
    // and uses it to see if any system messages were entered.
    switch(fw[0])
    {
        // Return the current time.
        case 'thetime':
        var date = new Date();
        var date_string = date.toString();
        append_line(date_string, 0, 0);
        break;
        // Echo some text.
        case 'echo':
        var echo_text = line.substring(5)
        if (echo_text){
            append_line(echo_text, 0, 0);
        }else{
            // For ducks with an empty echo.
            append_line("error: quack quack", 0, 0);
        }
        break;
        // Help menu, not finished (TODO).
        case 'help':
        var help_text = "haha no help for u"
        append_line(help_text, 0, 0);
        break;
        // This simply displays the current directory for now;
        // it's useless and will probably get removed. (TODO)
        case 'cd':
        var loc = window.location.pathname;
        var dir = loc.substring(0, loc.lastIndexOf('/'));
        append_line(dir, 0, 0);
        break;
        // Count the (alphanumeric) letters used in the chat
        // This could get removed or expanded as it might
        // or might not be useful for something. (TODO)
        case 'letters':
        var content = history.innerHTML;
        var letters = null;
        letters = content.replace(/<(?!br\s*\/?)[^>]+>/g, '');
        letters = letters.length;
        append_line(letters, 0, 0);
        break;
        // Clear the chat history box.
        case 'clear':
        history.innerHTML="";
        process('thetime');
        break;
        // Display your ip to the guy on the other side...
        // This could be useful for proving who you are to your friend
        // although it is probably quite dumb and needs something better
        // to replace it! (TODO)
        case 'ip':
        append_line(myip, 0, 0);
        break;
        default:
        break;
    }
    // split the possible filetype at the end of the URL.
    var filetype = line.split('.').pop();
    // Take out unnecessary text from the URL.
    var website = line.replace("http://", "").replace("https://", "").replace("www.", "");
    // Split website into an array by the period.
    website = website.split('.');
    // Embed certain image types into the line
    if (line){
        switch(filetype){
            case 'png':
            case 'jpg':
            case 'jpeg':
            case 'gif':
            case 'bmp':
            append_line("<img src=\""+line+"\">", 0, 0);
            console.log('attempting to parse image: '+line);
            break;
            default:
            break;
        }
        // This section parses YouTube videos into chat.
        // if [youtube]
        if (website[0] == "youtube"){
            var youtube_url = line.split("=").pop();
            append_line("<iframe width=\"420\" height=\"315\" src=\"http://www.youtube.com/embed/"+youtube_url+"\" frameborder=\"0\" allowfullscreen></iframe>", 0, 0);
            console.log('attempting to parse YouTube video: '+line);
        }
        // if [youtu] (for youtu.be URL)
        if (website[0] == "youtu"){
            var youtube_url = line.split("be/").pop();
            append_line("<iframe width=\"420\" height=\"315\" src=\"http://www.youtube.com/embed/"+youtube_url+"\" frameborder=\"0\" allowfullscreen></iframe>", 0, 0);
            console.log('attempting to parse YouTube video: '+line);
        }
    }
}

// this function pre-processes chat input
// returns new chat string which will be sent
// to other clients and displayed in the chat box
function pre_process(line) {
    emoticons_enabled = 1
    if (emoticons_enabled) line = emoticons_replace(line);
    return line;
}


public/moq_emoticons.js

// emoticon replacement dictionary
var emoticon_list = {
    ':)' : '｡◕‿◕｡',
    '(:' : '(▰˘◡˘▰)',
    ':p' : '（〜^∇^)〜',
    ';p' : '(-‿◦☀)',
    ';)' : '(。^_-)ノ',
    ':D' : 'o(〃＾▽＾〃)o',
    ';D' : '(●´∀｀●)',
    '._.' : '(^._.^)ﾉ',
    '.__.' : '｡◕‿‿◕｡',
    '._.;' : '(,,◕　⋏　◕,,)',
    '._.\'' : '(,,◕　⋏　◕,,)',
    '._.!' : '(づ｡◕‿‿◕｡)づ',
    ':o' : '(゜。゜)',
    ':O' : '(´⊙ω⊙`)！',
    'o_o' : '＼(○o○)ノ',
    '><' : 'ヽ(≧Д≦)ノ',
    '>_<' : '（＞д＜）',
    '>.<' : '(*￣m￣)',
    ':B' : '(●￣(ｴ)￣●)',
    '^o^' : '(ノ^o^)ノ',
    '<_<' : 'ヘ(￣ω￣ヘ)',
    '<.<' : '〜(￣△￣〜)',
    '>_>' : '(ﾉ≧∀≦)ﾉ',
    '>.>' : '(ﾉ◕ヮ◕)ﾉ*:・ﾟ✧',
    ':x' : '(✖╭╮✖)',
    'x)' : '✖‿✖',
    'x_x' : '(=ｘェｘ=)',
    'u_u' : 'U￣ｰ￣U',
    '\\o/' : '＼（＠￣∇￣＠）／',
    '^^!' : '☆*･゜ﾟ･*\\(^O^)/*･゜ﾟ･*☆',
    ':/' : '（ー△ー；）',
    ':\\' : '(一。一;;）',
    'u_u' : '(▰︶︹︺▰)',
    'u3u' : '(︶ω︶)',
    '>3<' : 'о(ж＞▽＜)ｙ ☆',
    'n_n' : '(✿◠‿◠)',
    'n_n\'' : 'ヾ（*⌒ヮ⌒*）ゞ',
    ':n' : '(っ˘̩╭╮˘̩)っ',
    'n_N' : '｢(＝＞o≦＝)ﾉ',
    '<3' : '(≚ᄌ≚)ℒℴѵℯ❤',
    'E>' : '(｡♥‿♥｡)',
    ':v' : '（＾ｖ＾）',
    'XD' : '(ノ＞▽＜。)ノ',
    'xD' : 'o(≧∇≦o)',
    'xP' : 'p(≧∇≦)q',
    '^^' : '(。⌒∇⌒)。',
    '^_^' : '(=^-ω-^=)',
    '^.^' : '｡^‿^｡',
    '^-^' : '（‐＾▽＾‐）',
    ';_;' : '.・゜゜・（／。＼）・゜゜・．',
    ';.;' : '｡：ﾟ(｡ﾉω＼｡)ﾟ･｡',
    ';__;' : '(இ﹏இ`｡)',
    '; ;' : '｡ﾟ( ﾟஇ‸இﾟ)ﾟ｡',
    'T_T' : 'o(╥﹏╥)o',
    'T__T' : '(━┳━ _ ━┳━)',
    'T.T' : '(。┰ω┰。)',
    'T-T' : '｡：ﾟ(｡ﾉω＼｡)ﾟ･｡',
    ':3' : '｡＾⋏＾｡',
    ':(' : '(▰˘︹˘▰)',
    ';(' : '(っ˘̩╭╮˘̩)っ',
    'o/' : 'ヾ(＠⌒ー⌒＠)ノ',
    'z_z' : '(∪｡∪)｡｡｡zzz',
    'z.z' : '(≚ᄌ≚)ƶƵ'
}

// replaces any basic emoticons in the given string with more interesting ones
function emoticons_replace(line) {
    var re = /(?:^|\s)(\S(?:_*|\s)\S{1,2})(?=\s|$)/g; // match potential emoticons
    var result = line;
    var m;
    while ((m = re.exec(line)) != null) {
        if (emoticon_list[m[1]]) {
            // matched emoticon should be replaced
            result = result.replace(m[1], emoticon_list[m[1]]);
        }
        else {
            // check if unextended version of matched
            // emoticon should be replaced
            if (m[1][1] == '_' && m[1][m[1].length-1] != '_' &&
                m[1][m[1].length-2] == '_') {
                // check for most extended valid emoticons
                if (emoticon_list[m[1][0] + '__' + m[1][m[1].length-1]]) {
                    result = result.replace(m[1], emoticon_list[m[1][0] +
                        '__' + m[1][m[1].length-1]] +
                            (Array(m[1].length - 3).join("〜")));
                }
                // check for least extended valid emoticons
                else if (emoticon_list[m[1][0] + '_' + m[1][m[1].length-1]]) {
                    result = result.replace(m[1], emoticon_list[m[1][0] +
                        '_' + m[1][m[1].length-1]] +
                            (Array(m[1].length - 3).join("〜")));
                }
            }
            // normal text
        }
    }
    return result;
}


public/chat.js
console.log('moQ chat processor');

// Connect socket
  var socket = io.connect('http://localhost:3000');
  socket.on('news', function (data) {
    socket.emit('my other event', { my: 'data' });
  });

// Put the version number here.
document.getElementById('verid').innerHTML="0.01";

// History is the chat log area.
// Textinput is the box that you type in to.
// Displayname is the client username.
// Submitbutton is the submit button from the input area.
var history = document.getElementById('history');
var textinput = document.getElementById('textinput');
var displayname = document.getElementById('displayname');
var submitbutton = document.getElementById('submitbutton');

var data = "";

// Focus the caret to the text input box, so the user can type right away!
textinput.focus();

// Clear the history box, just incase?
history.innerHTML="";

// Run through vars and update default values to stored ones
document.getElementById('displayname').innerHTML=localStorage.displayname;

// This function takes the line that the user has typed in the input box
// and places it into the chat history box!
function append_line(line, system){
    if (line.length){ 
        // There are user messages and system messages, which can be styled
        // and differentiated in the css.
        if(system == 1){
            // New content is added using appendChild, as content does not
            // have to be deleted and reloaded this way.
            var new_content = document.createElement('span');
            new_content.innerHTML = "<span class=\"displayname\">"+localStorage.displayname+"</span><span class=\"user\">"+line+"</span>";
            history.appendChild(new_content);
        }else if(system == 0){
            var new_content = document.createElement('span');
            new_content.innerHTML = "<span class=\"system\">"+line+"</span>";
            history.appendChild(new_content);
        }else if(system == 2){
            // This is for appending multiple lines for the user message.
            var new_content = document.createElement('span');
            new_content.innerHTML = "<span class=\"user\">"+line+"</span>";
            history.appendChild(new_content);    
        }else if(system == 3){
            // This is a hidden message.
            return;
        }
    }

    // Save the chat log into storage.
    // This is done every time a line is appended.
    localStorage.log=history.innerHTML;

    // Scroll to the bottom of the history when a new message is appended.
    history.scrollTop = history.scrollHeight;
}

// this gets called whenever the user presses something on their
// keyboard. 
function run(e) {
    // This code might be useful for responsive design or mobile layout
    // when I get around to doing that... (TODO)
    //if (window.innerHeight <= 900){
        // history.fontSize="100%";
        // textinput.fontSize="100%";
    //}else{
    //}
    switch(e.keyCode)
    {
        // When enter button is pressed, append and process the line
        // and reset the input box to empty.
        case 13:
        e.preventDefault();
        send_text();
        break;
        // Disable up arrow key (interferes with input box).
        case 38:
        e.preventDefault();
        break;
        // Disable down arrow key (interferes with input box).
        case 40:
        e.preventDefault();
        break;
        default:
        // If text message greater than X characters, disable input.
        // ctrl f 'max message size' for explanation.
        if (textinput.value.length >= 255){
            e.preventDefault();
        }
        break;
    }

}

// This function sends text from the input box to the history box
// It is used by enter key listener and the submit button.
function send_text(){
    data = textinput.value;
    // Send the value of the text to the server.
    socket.emit('msg_send', data);
    // Place the value of the text input box to the history box.
    append_line(data, 1);
    // Run the value through the processor.
    process(data);
    // Reset the text input box to a blank state.
    textinput.value="";
}

// When a new message is received...
socket.on('msg_new', function(data){
    // Make sure the data is not empty.
    if(data.length){
        // Append the data to the chat history.
        append_line(data, 1);
        // Process the data.
        process(data);
        // Check to make sure this is getting called correctly
        console.log("received data: "+data);
    }
});

// This function processes each line as it is appended,
// which is used for the system messages and most likely
// for embedding content (when that functionality is added).
// (TODO)
// I also want to be able to have the commands only locally
// displayed, for example if someone types clear, it should
// not be sent to everyone as a text message.
// (TODO)
function process(line){
    // Split the lines by the spaces
    var fw = line.split(" ");
    // This switch case selects the first word of the split
    // and uses it to see if any system messages were entered.
    switch(fw[0])
    {
        // Return the current time.
        case 'thetime':
        var date = new Date();
        var date_string = date.toString();
        append_line(date_string, 0);
        break;
        // Echo some text.
        case 'echo':
        var echo_text = line.substring(5)
        if (echo_text){
            append_line(echo_text, 0);
        }else{
            // For ducks with an empty echo.
            append_line("error: quack quack", 0);
        }
        break;
        // Help menu, not finished (TODO).
        case 'help':
        var help_text = "haha no help for u"
        append_line(help_text, 0);
        break;
        // This simply displays the current directory for now;
        // it's useless and will probably get removed. (TODO)
        case 'cd':
        var loc = window.location.pathname;
        var dir = loc.substring(0, loc.lastIndexOf('/'));
        append_line(dir, 0);
        break;
        // Count the (alphanumeric) letters used in the chat
        // This could get removed or expanded as it might
        // or might not be useful for something. (TODO)
        case 'letters':
        var content = history.innerHTML;
        var letters = null;
        letters = content.replace(/<(?!br\s*\/?)[^>]+>/g, '');
        letters = letters.length;
        append_line(letters, 0);
        break;
        // Clear the chat history box.
        case 'clear':
        history.innerHTML="";
        process('thetime');
        break;
        // Display your ip to the guy on the other side...
        // This could be useful for proving who you are to your friend
        // although it is probably quite dumb and needs something better
        // to replace it! (TODO)
        case 'ip':
        append_line(myip, 0);
        break;
        default:
        break;
    }

    // split the possible filetype at the end of the URL.
    var filetype = line.split('.').pop();
    // Take out unnecessary text from the URL.
    var website = line.replace("http://", "").replace("https://", "").replace("www.", "");
    // Split website into an array by the period.
    website = website.split('.');

    // Embed certain image types into the line
    if (line){
        switch(filetype){
            case 'png':
            case 'jpg':
            case 'jpeg':
            case 'gif':
            case 'bmp':
            append_line("<img src=\""+line+"\">", 0);
            console.log('attempting to parse image: '+line);
            break;
            default:
            break;
        }
        // This section parses YouTube videos into chat.
        // if [youtube]
        if (website[0] == "youtube"){
            var youtube_url = line.split("=").pop();
            append_line("<iframe width=\"420\" height=\"315\" src=\"http://www.youtube.com/embed/"+youtube_url+"\" frameborder=\"0\" allowfullscreen></iframe>", 0);
            console.log('attempting to parse YouTube video: '+line);
        }
        // if [youtu] (for youtu.be URL)
        if (website[0] == "youtu"){
            var youtube_url = line.split("be/").pop();
            append_line("<iframe width=\"420\" height=\"315\" src=\"http://www.youtube.com/embed/"+youtube_url+"\" frameborder=\"0\" allowfullscreen></iframe>", 0);
            console.log('attempting to parse YouTube video: '+line);
        }
    }
}

// This function stores the chat in localStorage, so that
// the chat history is retained when the page is refreshed.
// It needs to separate the chat logs by what rooms they are in
// and probably needs some server request to make sure that
// the data was not lost - this still has a lot of work to be done.
// (TODO)
function text_log()
{
    // Make sure there is storage
    if(typeof(Storage)!=="undefined"){
        if (localStorage.log){
            // Put history contents back in its' DIV.
            history.innerHTML=localStorage.log;
        }
        else
        {
            // If it is a brand new chat - this message is displayed.
            localStorage.log="<span class=\"system\">New chat with %user%</span>";
        }
        // Do we have a display name stored?
        if(localStorage.displayname){
            // Set the display name to the one in storage.
            displayname.innerHTML=localStorage.displayname;
        }else{
            // They don't have a name.
            localStorage.displayname="Unnamed";
        }
    }else{
        // Browser does not support localStorage. A server request would
        // be done here, possibly. Server could store the logs in a 
        // database, if this things turns out that way. (TODO)
        history.innerHTML="Logs could not be loaded with this browser.";
    }
}
// Execute the text log function on page load.
text_log()

// This function stores the settings when the user updates them
// and also updates setting specific things in moQ.
function update_settings(){
    // Set the username to the one we have in storage (if it exists).
    if(typeof(Storage)!=="undefined"){
        if (localStorage.displayname){
            localStorage.displayname=displayname.innerHTML;

        // Also set the display name to chat history.
        document.getElementById('displayname').innerHTML=localStorage.displayname;
    }
}

    // Every time settings is updated, go through these colours and update them.
    document.getElementById('dncolour').style.color=document.getElementById('dncolour').innerHTML;
    document.getElementById('displayname').style.color=document.getElementById('dncolour').innerHTML;
    document.getElementById('textcolour').style.color=document.getElementById('textcolour').innerHTML;

    // Select every element with user class.
    var user_colour = document.getElementsByClassName('user');

    // Go through every element with the class and set the text colour to the one in settings.
    for (var i=user_colour.length; i--;) {
        user_colour[i].style.color=document.getElementById('textcolour').innerHTML;
    }

    // Select every element with displayname class.
    var displayname_color = document.getElementsByClassName('displayname');

    // Go through every element with the class and set the text colour to the one in settings.
    for (var i=displayname_color.length; i--;) {
        displayname_color[i].style.color=document.getElementById('dncolour').innerHTML;
    }
}

// When the settings button is pressed, this function toggles the
// settings menu from the top.
function settings_toggle(){
    if(document.getElementById('settings').style.top != "50%"){
        document.getElementById('settings').style.top="50%";
    }else{
        document.getElementById('settings').style.top="-100%";
    }
}

// This function captures any clicks made and redirects focus to
// the text input box. Users can still select text and click links
// and stuff, this just makes it less hassle to type stuff.
document.addEventListener('click', function(e) {
    // At the moment, it takes focus when it should not, for
    // example when the user is entering things in the settings
    // page. This needs to be fixed. (TODO)
    //textinput.focus();
}, false);

// Start with the history scrolled to the bottom.
setTimeout(function(){history.scrollTop = history.scrollHeight;},200);
